nestedtensor/csrc/nested_tensor_impl.h

```diff
@@ -0,0 +1,75 @@
+#pragma once
+
+#include <ATen/ATen.h>
+#include <nestedtensor/csrc/nested_tensor.h>
+
+namespace at {
+
+constexpr auto NestedTensorKey = DispatchKey::PrivateUse1_PreAutograd;
+
+struct NestedTensorImpl : public c10::TensorImpl {
+  explicit NestedTensorImpl(torch::nested_tensor::NestedTensor&& data)
+      : TensorImpl(
+            c10::DispatchKeySet(NestedTensorKey),
+            data.dtype(),
+            data.device()),
+        _data(std::move(data)) {}
+
+  int64_t dim() const override {
+    return _data.dim();
+  }
+  int64_t numel() const override {
+    return _data.numel();
+  }
+  bool is_contiguous(
+      at::MemoryFormat memory_format) const override {
+    return _data.is_contiguous();
+  }
+
+  IntArrayRef sizes() const override;
+  int64_t size(int64_t dim) const override;
+  IntArrayRef strides() const override;
+
+  torch::nested_tensor::NestedTensor _data;
+
+};
+
+inline torch::nested_tensor::NestedTensor get_nested_tensor(
+    const at::Tensor tensor) {
+  if (!tensor.unsafeGetTensorImpl()->key_set().has(at::NestedTensorKey)) {
+    throw std::runtime_error("Function requires NestedTensorImpl");
+  }
+  auto nt_impl =
+      static_cast<at::NestedTensorImpl*>(tensor.unsafeGetTensorImpl());
+  return nt_impl->_data;
+}
+
+inline at::NestedTensorImpl* get_nested_tensor_impl(const at::Tensor tensor) {
+  if (!tensor.unsafeGetTensorImpl()->key_set().has(at::NestedTensorKey)) {
+    throw std::runtime_error("Function requires NestedTensorImpl");
+  }
+  return static_cast<at::NestedTensorImpl*>(tensor.unsafeGetTensorImpl());
+}
+
+inline at::Tensor wrap_nested_tensor(
+    torch::nested_tensor::NestedTensor&& result) {
+  return at::detail::make_tensor<NestedTensorImpl>(std::move(result));
+}
+
+inline at::Tensor wrap_tensor_node(
+    torch::nested_tensor::TensorNode&& result) {
+  return at::detail::make_tensor<NestedTensorImpl>(
+      torch::nested_tensor::NestedTensor(std::move(result)));
+}
+
+Tensor NestedTensor_to_tensor(Tensor tensor, c10::optional<int64_t> dim_);
+
+inline std::ostream& operator<<(std::ostream& out, const NestedTensorImpl& batch_tensor) {
+  auto node = batch_tensor._data.get_structure();
+  out << "NESTED_TENSOR";
```

Comment:
nit: remove?

Reply:
This is useful for debugging when we want to print an instance of the impl, but, yes, maybe not necessary if we want to reduce the code a bit